pyro/nn/auto_reg_nn.py

```diff
@@ -7,6 +7,70 @@
 from torch.nn import functional as F
 
 
+def sample_mask_indices(input_dim, hidden_dim, simple=True):
+    """
+    Samples the indices assigned to hidden units during the construction of MADE masks
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer
+    :type hidden_dim: int
+    :param simple: True to space fractional indices by rounding to nearest int, false round randomly
+    :type simple: bool
+    """
+    indices = torch.linspace(1, input_dim, steps=hidden_dim)
+    if simple:
+        # Simple procedure tries to space fractional indices evenly by rounding to nearest int
+        return torch.round(indices)
+    else:
+        # "Non-simple" procedure creates fractional indices evenly then rounds at random
+        ints = indices.floor()
+        ints += torch.bernoulli(indices - ints)
+        return ints
+
+
+def create_mask(input_dim, observed_dim, hidden_dim, num_layers, permutation, output_dim_multiplier):
+    """
+    Creates MADE masks for a conditional distribution
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param observed_dim: the dimensionality of the variable that is conditioned on (for conditional densities)
+    :type observed_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer(s)
+    :type hidden_dim: int
+    :param num_layers: the number of hidden layers for which to create masks
+    :type num_layers: int
+    :param permutation: the order of the input variables
+    :type permutation: torch.Tensor
+    :param output_dim_multiplier: tiles the output (e.g. for when a separate mean and scale parameter are desired)
+    :type output_dim_multiplier: int
+    """
+    # Create mask indices for input, hidden layers, and final layer
+    # We use 0 to refer to the elements of the variable being conditioned on,
+    # and range(1:(D_latent+1)) for the input variable
+    var_index = permutation.clone().type(torch.get_default_dtype())
+    var_index[permutation] = torch.arange(input_dim, dtype=torch.get_default_dtype())
+
+    # Create the indices that are assigned to the neurons
+    input_indices = torch.cat((torch.zeros(observed_dim), 1 + var_index))
+    hidden_indices = [sample_mask_indices(input_dim, hidden_dim) for i in range(num_layers)]
+    output_indices = (var_index + 1).repeat(output_dim_multiplier)
+
+    # Create mask from input to output for the skips connections
+    mask_skip = (output_indices.unsqueeze(-1) > input_indices.unsqueeze(0)).type_as(output_indices)
```

Comment:
I suppose we can replace all the `.type_as` with `.float()` instead. That will directly pick up the default tensor type whichever is used.